realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -859,21 +861,219 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
     return -1;
 }
 
+std::unique_ptr<Query> getHandoverQuery (jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr) {
+    SharedGroup::Handover <Query> *handoverQueryPtr = reinterpret_cast<SharedGroup::Handover <Query> * > (queryPtr);
+    std::unique_ptr <SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
+
+    SG(bgSharedGroupPtr)->end_read();
+
+    SharedGroup::VersionID currentVersion = SG(bgSharedGroupPtr)->get_version_of_current_transaction();
+    bool isDifferentVersions = (currentVersion != handoverQuery->version);
+    if (isDifferentVersions) {
+        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
+    } else {
+        SG(bgSharedGroupPtr)->begin_read();
+    }
+
+    std::unique_ptr <Query> query = SG(bgSharedGroupPtr)->import_from_handover(std::move(handoverQuery));
+    if (isDifferentVersions) {
+        LangBindHelper::advance_read(*SG(bgSharedGroupPtr), *CH(replicationPtr));
```

Comment:
the catch is in the caller method ex `Java_io_realm_internal_TableQuery_nativeFindWithHandover`